### ISSUE TITLE
Cache configuration

### ISSUE THREAD

**Issue by rdegges** [stormpath/stormpath-sdk-spec#16]
The existing cacheManager stuff is really... vague. EG: How does it look if a user wants us to cache to a Redis box? A memcached box? A cluster of memcached servers?

We should have this documented in here somewhere, so that all integrations can properly reflect the same usage. This means passing in strings for hosts / etc., not passing in fully constructed cache objects, etc.

**Comment by lhazlewood** [comment]
I don't think configuration of 3rd party services has a place in our specification at all.

In the Java world, you have a `CacheManager` interface - as long as anyone implements that interface to plug in to whatever cache mechanism they prefer, our SDK 'just works'.  Trying to figure out all the different config options for Redis vs Hazelcast vs Memcache is a concern totally orthogonal to the SDK and our integrations and is better configured separately.

**Comment by rdegges** [comment]
Hmm... I disagree.

I think we should definitely allow it to be configured, as it directly affects the developer using our library.

For instance, not all memcached/redis clients support the exact same method names / params, so allowing a user to pass in some 'arbitrary' cache object seems like a hack to me.

I think it'd be cool to allow for some basic params to be set, something like:

``` json
{
  "host": "host_string_here (optional - default: localhost)",
  "username": "xxx (optional)",
  "password": "xxx (optional)",
  "port": 1,  // optional, default depends on type set
  "type": "redis", // default: "memory"
}
```

**Comment by lhazlewood** [comment]
My point is that memcache or redit should sure as heck have their own way of configuring these things - why is it our responsibility?  Let's just use what they already have...

**Comment by lhazlewood** [comment]
https://github.com/NodeRedis/node_redis
https://github.com/3rd-Eden/memcached
(I'm sure there are a few more)

We shouldn't be responsible for making yaml config work for 3rd party libraries.  First, you don't know which client/library/framework the app developer wants to use, and 2nd, whatever they choose, they should configure it according to that project's instructions/documentation.

All we need to communicate is how to 'plug in' those 3rd party things into our SDK/integration.

**Comment by robertjd** [comment]
I agree, our spec should not define how to configure 3rd party libraries.  It should define the parameters for our cache manager, but the actual interface to the cache (a redis client, a memcached client) should be passed by the developer, so that they configure it as they see fit.  The SDK has to inspect the incoming instance and choose the right adapter.

**Comment by rdegges** [comment]
I disagree still. I mean, take Django for instance, here's their official settings documentation: https://docs.djangoproject.com/en/1.9/ref/settings/#caches

Here's the ones for Laravel: https://laravel.com/docs/5.1/cache

Same thing with Rails: http://guides.rubyonrails.org/caching_with_rails.html

They go out of the way to provide their own adapters to the caches they support, and let all developers using django specify some simple options for setting things up, without requiring them to know anything at all about some caching libraries.

IMO, this is a matter of DX. Do we want to go the extra mile to make it a bit easier? Most frameworks offer this functionality, which makes me thing that since we are (essentially) a security framework, we should too!

**Comment by robertjd** [comment]
I've read through the linked Django docs, and here is what I see:

They have the same concept of a cache manager, like we have.  This is what takes the key prefix, timeout, and related configuration options.  Internally, Django has implemented some adapters between their cache manager and cache implementations, such as memcached.  They also allow you to write your own adapter and specify the class path.  This is all great and awesome.

From the documentation, it appears that Django will create a client instance of the cache implementation for me.  I declare which implementation I would like to use with the `BACKEND` option.  If I need to configure it, e.g. configure the memcached client with a custom port, I have to use the ["options" paramater](https://docs.djangoproject.com/en/1.9/ref/settings/#options), for which they state:

> Extra parameters to pass to the cache backend. Available parameters vary depending on your cache backend.

As such, they are taking the approach that Les and I are arguing for: they aren't going to couple directly to the options for the 3rd party cache implementation, as that that coupling requires maintenance.  They simply give you a parameter that allows you to pass all that through to the client constructor.

@rdegges where is the problem with this approach?

**Comment by nbarbettini** [comment]
To add on to Robert's point, I think that the DX can be improved by simply having documentation around how to configure integration X with cache Y - without necessarily coupling us directly to a 3rd party implementation.
